freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingService.java

```diff
@@ -0,0 +1,122 @@
+        validationService.validateStackForDownscale(allInstances, stack, new ScalingPath(originalFormFactor, request.getTargetFormFactor()));
+        return triggerDownscale(request, stack, originalFormFactor);
+    }
+
+    private Stack getStack(String accountId, String envCrn) {
+        Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(envCrn, accountId);
+        MDCBuilder.buildMdcContext(stack);
+        return stack;
+    }
+
+    private UpscaleResponse triggerUpscale(UpscaleRequest request, Stack stack, FormFactor originalFormFactor) {
+        Operation operation = startScalingOperation(stack.getAccountId(), request.getEnvironmentCrn(), OperationType.UPSCALE);
+        UpscaleEvent upscaleEvent = new UpscaleEvent(UpscaleFlowEvent.UPSCALE_EVENT.event(),
+                stack.getId(), request.getTargetFormFactor().getInstanceCount(), false, false, false, operation.getOperationId());
+        LOGGER.info("Trigger upscale flow with event: {}", upscaleEvent);
+        FlowIdentifier flowIdentifier = flowManager.notify(UpscaleFlowEvent.UPSCALE_EVENT.event(), upscaleEvent);
+        UpscaleResponse response = new UpscaleResponse();
+        response.setOperationId(operation.getOperationId());
+        response.setOriginalFormFactor(originalFormFactor);
+        response.setTargetFormFactor(request.getTargetFormFactor());
+        response.setFlowIdentifier(flowIdentifier);
+        return response;
+    }
+
+    private DownscaleResponse triggerDownscale(DownscaleRequest request, Stack stack, FormFactor originalFormFactor) {
+        Operation operation = startScalingOperation(stack.getAccountId(), request.getEnvironmentCrn(), OperationType.DOWNSCALE);
+        ArrayList<String> instanceIdList = getDownscaleCandidates(stack, originalFormFactor, request.getTargetFormFactor());
+        DownscaleEvent downscaleEvent = new DownscaleEvent(DownscaleFlowEvent.DOWNSCALE_EVENT.event(),
+                stack.getId(), instanceIdList, request.getTargetFormFactor().getInstanceCount(), false, false, false, operation.getOperationId());
+        LOGGER.info("Trigger downscale flow with event: {}", downscaleEvent);
+        FlowIdentifier flowIdentifier = flowManager.notify(DownscaleFlowEvent.DOWNSCALE_EVENT.event(), downscaleEvent);
+        DownscaleResponse response = new DownscaleResponse();
+        response.setOperationId(operation.getOperationId());
+        response.setOriginalFormFactor(originalFormFactor);
+        response.setTargetFormFactor(request.getTargetFormFactor());
+        response.setFlowIdentifier(flowIdentifier);
+        return response;
+    }
+
+    private ArrayList<String> getDownscaleCandidates(Stack stack, FormFactor originalFormFactor, FormFactor targetFormFactor) {
+        int instancesToRemove = originalFormFactor.getInstanceCount() - targetFormFactor.getInstanceCount();
+        return stack.getNotDeletedInstanceMetaDataSet().stream()
```

Comment:
it would be nice if non healthy/available instances would be preferred over healthy

Reply:
we explicitly validate for any unhealthy instances beforein `isAnyInstanceInNotAvailableState` and do not let the scaling to proceed if such exists

Reply:
I know, but it would be more future proof if that logic changes